### PR TITLE
[SPARK-38895][SQL] Unify the shuffle read canonicalized

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

- Override the isCanonicalizedPlan in AQEShuffleRead
- cleanup unnecessary code

### Why are the changes needed?

After canonicalized, the child of AQEShuffleReadExec will be a exchange instead of shuffle query stage. For better maintenance, we can simply override the isCanonicalizedPlan and let famework to check if the plan can be executed.

### Does this PR introduce _any_ user-facing change?

no, jsut a small code refactor

### How was this patch tested?

Pass existed test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala [hunk 1]
```diff
@@ -144,12 +150,7 @@ case class AQEShuffleReadExec private(
     }
   }
 
-  private def shuffleStage = child match {
-    case stage: ShuffleQueryStageExec => Some(stage)
-    case _ => None
-  }
-
-  @transient private lazy val partitionDataSizes: Option[Seq[Long]] = {
+  private def partitionDataSizes: Option[Seq[Long]] = {
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:153]: only called once, not see a benefit to materialize it

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala [hunk 2]
```diff
@@ -67,8 +74,7 @@ case class AQEShuffleReadExec private(
             case e: Expression => r.updateAttr(e).asInstanceOf[Partitioning]
             case other => other
           }
-        case _ =>
-          throw new IllegalStateException("operating on canonicalization plan")
+        case _ => super.outputPartitioning
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:77]: it seems overkill to throw error and other canonicalized spark plan can also work

**Comment by ulysses-you** [comment]
cc @cloud-fan @maryannxue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala [hunk 3]
```diff
@@ -48,6 +48,13 @@ case class AQEShuffleReadExec private(
     assert(partitionSpecs.forall(_.isInstanceOf[PartialMapperPartitionSpec]))
   }
 
+  private def shuffleStage = child match {
+    case stage: ShuffleQueryStageExec => Some(stage)
+    case _ => None
+  }
+
+  override protected def isCanonicalizedPlan: Boolean = shuffleStage.isEmpty
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:56]: what's wrong with the default implementation of `isCanonicalizedPlan`?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:56]: what's wrong with the default implementation of `isCanonicalizedPlan`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:153]: can we add `assert(shuffleStage.isDefined)` like what we did in `lazy val shuffleRdd`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala [hunk 4]
```diff
@@ -207,47 +208,38 @@ case class AQEShuffleReadExec private(
   }
 
   @transient override lazy val metrics: Map[String, SQLMetric] = {
-    if (shuffleStage.isDefined) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:210]: can we add `assert(shuffleStage.isDefined)` like what we did in `lazy val shuffleRdd`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala]: The default implementation of isCanonicalizedPlan is correct. here we just make it clear: if the plan is cannonicalized then the shuffle stage is empty. I change to use `super.isCanonicalizedPlan && shuffleStage.isEmpty`

**Pull request closed without merging**